Project/Assets/ML-Agents/Examples/Match3/Scripts/Match3Agent.cs

```diff
@@ -0,0 +1,359 @@
+using System;
+using UnityEngine;
+using Unity.MLAgents;
+using Unity.MLAgents.Actuators;
+using Unity.MLAgents.Extensions.Match3;
+
+namespace Unity.MLAgentsExamples
+{
+
+    /// <summary>
+    /// State of the "game" when showing all steps of the simulation. This is only used outside of training.
+    /// The state diagram is
+    ///
+    ///      | <--------------------------------------- ^
+    ///      |                                          |
+    ///      v                                          |
+    ///  +--------+      +-------+      +-----+      +------+
+    ///  |Find    | ---> |Clear  | ---> |Drop | ---> |Fill  |
+    ///  |Matches |      |Matched|      |     |      |Empty |
+    ///  +--------+      +-------+      +-----+      +------+
+    ///
+    ///    |     ^
+    ///    |     |
+    ///    v     |
+    ///
+    ///  +--------+
+    ///  |Wait for|
+    ///  |Move    |
+    ///  +--------+
+    ///
+    /// The stats advances each "MoveTime" seconds.
+    /// </summary>
+    enum State
+    {
+        /// <summary>
+        /// Guard value, should never happen.
+        /// </summary>
+        Invalid = -1,
+
+        /// <summary>
+        /// Look for matches. If there are matches, the next state is ClearMatched, otherwise WaitForMove.
+        /// </summary>
+        FindMatches = 0,
+
+        /// <summary>
+        /// Remove matched cells and replace them with a placeholder value.
+        /// </summary>
+        ClearMatched = 1,
+
+        /// <summary>
+        /// Move cells "down" to fill empty space.
+        /// </summary>
+        Drop = 2,
+
+        /// <summary>
+        /// Replace empty cells with new random values.
+        /// </summary>
+        FillEmpty = 3,
+
+        /// <summary>
+        /// Request a move from the Agent.
+        /// </summary>
+        WaitForMove = 4,
+    }
+
+    public class Match3Agent : Agent
+    {
+        [HideInInspector]
+        public Match3Board Board;
+
+        public float MoveTime = 1.0f;
+        public int MaxMoves = 500;
+
+        [Tooltip("If selected, will pick the move that makes the most points. Otherwise, will pick a random valid move.")]
+        public bool UseSmartHeuristic = false;
```

Comment:
Rename UseSmartHeuristic to SmartHeuristic. The word `Use` makes it sound like Heuristic will be used, but this is dependent on the Actuator setting "Force Heuristic".
Ideally, the `UseSmartHeuristic` checkbox should be next to `Force Heuristic` in the Actuator.
Also, I think it might confuse users that we now have 3 different places for the Heuristic:
Behavior Parameters (behavior type); Match3Agent(UseSmartHeuristic); and Actuator (ForceHeurisitic)
Not sure what can be done though...

Reply:
Changed to an enum for Greedy vs RandomValidMove, and updated some GameObject and behavior names.

> Ideally, the UseSmartHeuristic checkbox should be next to Force Heuristic in the Actuator.

Maybe for this specific case, but Greedy vs RandomValidMove is specific to the Match3Agent